addons/delivery/models/sale_order.py

```diff
@@ -29,6 +29,8 @@ def _compute_amount_total_without_delivery(self):
     def _compute_delivery_state(self):
         for order in self:
             order.delivery_set = any(line.is_delivery for line in order.order_line)
+            if order.delivery_set and order.carrier_id.delivery_type == 'onsite' and order.carrier_id.warehouse_id:
```

Comment:
It looks like `carrier_id.warehouse_id` is only defined as a field for `delivery.carrier` in `website_sale_picking`, so you should add this logic to that module. 

Also you should be setting `warehouse_id` with `_compute_warehouse_id` instead of `_compute_delivery_state` (i.e. override the existing compute). You should maybe check with the @odoo/rd-ecommerce  team if we should be doing it though because they added in a `TODO` to fix it: https://github.com/odoo/odoo/blob/aaeebb7fc89a0ee5f124b9749115b918d4815a24/addons/website_sale_picking/controllers/main.py#L29-L31

Reply:
I don't mind if you do fix it, it's a generic remark for future improvements of this logic.

Some things come to my mind though:
1) I'm not sure we want to recompute the `warehouse_id` field whenever a carrier is set, as it would replace the existing configuration, especially for backend orders.  I did write the TODO, but now I'm not sure that magic warehouse field reset would be appreciated by customers.  Overriding the methods update the carrier could be safer :thinking: 
2) Do we really consider 'Onsite Picking' as a generic carrier ? I'd say that it's only an eCommerce feature that doesn't have much meaning backend-wise (I believe using a fixed price carrier, and removing the 'Online payment' setting on the order gives the same behavior).

Reply:
1. I agree, which is why maybe it would make more sense if we do nothing for now and leave the implementation to you since it seems like more of a sales decision than a stock one in the end
2. For me it does make sense. Originally we were going to add an "Onsite Pickup" carrier into delivery and users would make one carrier per pickup location (i.e. different addresses/instructions for each one), but then it was added into ecommerce before we finished so we left it as is. To me it's wrong that the `warehouse_id` can only be set in delivery carriers when ecommerce is installed because someone may want to have onsite pickups without ecommerce installed (e.g. through a traditional SO with delivery carrier selected in the backend).

So I propose that either we (logistics) skip this part of our spec and submit it as feedback for your PO or we can add in a `compute_warehouse_id` override to `sale.order` within `website_sale_picking` so that it matches the behavior that `_validate_transaction_for_order` forces when the SO is validated through ecommerce (i.e. if carrier_id set + it is `onsite` delivery type and there's a warehouse_id is set, then updated it in this case, otherwise leave existing warehouse if set, else do the default behavior). It will just be annoying if the user specifies a warehouse and it's wiped in this case.

Reply:
Discussed with THD, dropped this spec from this task